File: registration_system_rs/src/activity/domain/activity.rs
use chrono::NaiveDateTime;

/// repository `update_activity` 使用的部分更新字段（None = 不修改）
#[derive(Debug, Default)]
pub struct UpdateActivityFields<'a> {
    pub name: Option<&'a str>,
    pub cover: Option<Option<&'a str>>,
    pub start_time: Option<NaiveDateTime>,
    pub end_time: Option<NaiveDateTime>,
    pub holding_date: Option<NaiveDateTime>,
    pub location: Option<&'a str>,
    pub location_latitude: Option<Option<f64>>,
    pub location_longitude: Option<Option<f64>>,
    pub opposing: Option<Option<&'a str>>,
    pub description: Option<Option<&'a str>>,
    pub home_team_id: Option<Option<i64>>,
    pub away_team_id: Option<Option<i64>>,
    pub color: Option<Option<&'a str>>,
    pub opposing_color: Option<Option<&'a str>>,
    pub players_per_team: Option<Option<i32>>,
    pub match_kind: Option<&'a str>,
    pub source_activity_id: Option<Option<&'a str>>,
    pub team_registration_count: Option<Option<i32>>,
}

/// 活动下各报名状态人数（管理后台统计条）
#[derive(Debug, Clone)]
pub struct RegistrationStandCounts {
    pub total: i64,
    pub unknown: i64,
    pub attending: i64,
    pub leave: i64,
    pub absent: i64,
}

/// 活动列表各状态数量（全库统计，不受分页筛选影响）
#[derive(Debug, Clone)]
pub struct ActivityStatusCounts {
    pub total: i64,
    pub registering: i64,
    pub ongoing: i64,
    pub ended: i64,
    pub cancelled: i64,
}

/// 分页活动列表（管理后台）
#[derive(Debug, Clone)]
pub struct ActivityListPage {
    pub items: Vec<Activity>,
    pub total: i64,
    pub page: u32,
    pub page_size: u32,
    pub counts: ActivityStatusCounts,
}

/// 分页报名列表（管理后台）
#[derive(Debug, Clone)]
pub struct RegistrationListPage {
    pub items: Vec<RegistrationWithInfo>,
    pub total: i64,
    pub page: u32,
    pub page_size: u32,
    pub counts: RegistrationStandCounts,
}

/// 报名记录 + 球员基本信息（管理后台用）
#[derive(Debug, Clone)]
pub struct RegistrationWithInfo {
    pub user_id: i64,
    pub stand: i8,
    pub registration_count: i32,
    pub paid: i8,
    pub operation_time: NaiveDateTime,
    pub checked_in_at: Option<NaiveDateTime>,
    pub checkin_distance_meters: Option<i32>,
    pub nickname: String,
    pub real_name: String,
    pub avatar_url: String,
    pub phone_number: String,
}

#[derive(Debug, Clone)]
pub struct Activity {
    pub id: String,
    pub cover: Option<String>,
    pub start_time: NaiveDateTime,
    pub end_time: NaiveDateTime,
    pub holding_date: NaiveDateTime,
    pub location: String,
    pub location_latitude: Option<f64>,
    pub location_longitude: Option<f64>,
    pub name: String,
    pub opposing: Option<String>,
    pub status: i8,
    pub description: Option<String>,
    pub home_team_id: Option<i64>,
    pub away_team_id: Option<i64>,
    pub color: Option<String>,
    pub opposing_color: Option<String>,
    pub players_per_team: Option<i32>,
    pub match_kind: Option<String>,
    pub source_activity_id: Option<String>,
    pub team_registration_count: Option<i32>,
    pub team_checkin_configs: Vec<ActivityTeamCheckInConfig>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, Clone)]
pub struct ActivityRegistration {
    pub id: i64,
    pub activity_id: String,
    pub user_id: i64,
    pub stand: i8,
    pub registration_count: i32,
    pub paid: i8,
    pub operation_time: NaiveDateTime,
    pub checked_in_at: Option<NaiveDateTime>,
    pub checkin_distance_meters: Option<i32>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, Clone)]
pub struct ActivityTeamCheckInConfig {
    pub activity_id: String,
    pub team_id: i64,
    pub enabled: bool,
    pub radius_meters: i32,
    pub open_minutes_before: i32,
    pub close_minutes_after: i32,
    pub updated_by_user_id: Option<i64>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

impl ActivityTeamCheckInConfig {
    pub fn checkin_open_at(&self, holding_date: NaiveDateTime) -> NaiveDateTime {
        holding_date - chrono::Duration::minutes(self.open_minutes_before as i64)
    }

    pub fn checkin_close_at(&self, holding_date: NaiveDateTime) -> NaiveDateTime {
        holding_date + chrono::Duration::minutes(self.close_minutes_after as i64)
    }
}

#[derive(Debug, Clone)]
pub struct ActivityCheckInRecord {
    pub id: i64,
    pub activity_id: String,
    pub team_id: i64,
    pub user_id: i64,
    pub latitude: f64,
    pub longitude: f64,
    pub distance_meters: i32,
    pub checked_in_at: NaiveDateTime,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

File: registration_system_backend_fe/src/services/activity.ts
import { http } from '@/utils/request'

export interface Activity {
  id: string
  name: string
  location: string
  location_latitude: number | null
  location_longitude: number | null
  status: number
  holding_date: string
  start_time: string
  end_time: string
  opposing: string | null
  cover: string | null
  description: string | null
  home_team_id: number | null
  away_team_id: number | null
  color: string | null
  opposing_color: string | null
  players_per_team: number | null
  match_kind: 'external' | 'internal' | null
  source_activity_id: string | null
  team_registration_count: number | null
  team_checkin_configs: ActivityTeamCheckinConfig[]
}

export interface ActivityTeamCheckinConfig {
  team_id: number
  enabled: boolean
  radius_meters: number
  open_minutes_before: number
  close_minutes_after: number
  checkin_open_at: string
  checkin_close_at: string
  updated_by_user_id: number | null
  updated_at: string
}

export interface RegistrationWithInfo {
  user_id: number
  stand: number
  stand_label: string
  registration_count: number
  paid: number
  operation_time: string
  checked_in_at: string | null
  checkin_distance_meters: number | null
  nickname: string
  real_name: string
  avatar_url: string
  phone_number: string
}

/** 各状态人数（全活动，不受当前筛选分页影响） */
export interface RegistrationStandCounts {
  total: number
  unknown: number
  attending: number
  leave: number
  absent: number
}

export interface RegistrationListPage {
  items: RegistrationWithInfo[]
  total: number
  page: number
  page_size: number
  counts: RegistrationStandCounts
}

/** 活动列表各状态数量（全库） */
export interface ActivityStatusCounts {
  total: number
  registering: number
  ongoing: number
  ended: number
  cancelled: number
}

export interface ActivityListPage {
  items: Activity[]
  total: number
  page: number
  page_size: number
  counts: ActivityStatusCounts
}

export interface CreateActivityPayload {
  name: string
  location: string
  location_latitude?: number
  location_longitude?: number
  holding_date: string
  start_time: string
  end_time: string
  opposing?: string
  cover?: string
  description?: string
  home_team_id?: number
  away_team_id?: number
  color?: string
  opposing_color?: string
  players_per_team?: number
}

export interface UpdateActivityPayload {
  name?: string
  location?: string
  location_latitude?: number | null
  location_longitude?: number | null
  holding_date?: string
  start_time?: string
  end_time?: string
  opposing?: string | null
  cover?: string | null
  description?: string | null
  home_team_id?: number | null
  away_team_id?: number | null
  color?: string | null
  opposing_color?: string | null
  players_per_team?: number | null
}

export interface UpdateActivityCheckinConfigPayload {
  team_id: number
  enabled: boolean
  radius_meters: number
  open_minutes_before: number
  close_minutes_after: number
}

export interface LocationSearchResult {
  provider_place_id: string
  title: string
  address: string
  display_name: string
  latitude: string
  longitude: string
}

export const STAND_LABELS: Record<number, string> = {
  0: '未表态',
  1: '参加',
  2: '请假',
  3: '迟到',
}

export const STAND_BADGE: Record<number, string> = {
  0: 'badge-ghost',
  1: 'badge-success',
  2: 'badge-warning',
  3: 'badge-error',
}

export const STATUS_LABEL: Record<number, string> = {
  0: '报名中',
  1: '进行中',
  2: '已结束',
  3: '已取消',
}

export const STATUS_BADGE: Record<number, string> = {
  0: 'badge-info',
  1: 'badge-success',
  2: 'badge-ghost',
  3: 'badge-error',
}

export interface ListActivitiesParams {
  page?: number
  page_size?: number
  /** -1 全部；0–3 按活动状态 */
  status?: number
  /** team = 有球队参与、可做球队内部报名的活动；direct = 无球队归属的直接活动 */
  registration_scope?: 'team' | 'direct'
}

export const listActivities = async (params?: ListActivitiesParams): Promise<ActivityListPage> => {
  return http.get<ActivityListPage>('/activities', { params })
}

export const searchActivityLocations = async (
  keyword: string,
  limit = 8,
): Promise<LocationSearchResult[]> => {
  return http.get<LocationSearchResult[]>('/activities/location-search', {
    params: { keyword, limit },
  })
}

export const resolveActivityLocation = async (
  latitude: number,
  longitude: number,
): Promise<LocationSearchResult> => {
  return http.get<LocationSearchResult>('/activities/location-resolve', {
    params: { latitude, longitude },
  })
}

export const getActivity = (id: string) => http.get<Activity>(`/activities/${id}`)

export const createActivity = (data: CreateActivityPayload) =>
  http.post<Activity>('/activities', data)

export const updateActivity = (id: string, data: UpdateActivityPayload) =>
  http.patch<void>(`/activities/${id}`, data)

export const updateActivityCheckinConfig = (id: string, data: UpdateActivityCheckinConfigPayload) =>
  http.patch<Activity>(`/activities/${id}/check-in-config`, data)

export const updateActivityStatus = (id: string, status: number) =>
  http.patch<void>(`/activities/${id}/status`, { status })

export const deleteActivities = (ids: string[]) =>
  http.delete<void>('/activities/batch', { data: { ids } })

export interface ListRegistrationsParams {
  page?: number
  page_size?: number
  /** -1 全部；0–3 按状态 */
  stand?: number
}

export const listRegistrations = async (
  activityId: string,
  params?: ListRegistrationsParams,
): Promise<RegistrationListPage> => {
  return http.get<RegistrationListPage>(`/activities/${activityId}/registrations`, { params })
}

export const adminRegister = (activityId: string, userId: number, stand: number) =>
  http.post<void>(`/activities/${activityId}/registrations`, {
    user_id: userId,
    stand,
    registration_count: stand === 1 ? 1 : 0,
  })

export const batchUpdateStand = (
  activityId: string,
  userIds: number[],
  stand: number,
  registrationCount?: number,
) =>
  http.patch<number>(`/activities/${activityId}/registrations/batch`, {
    user_ids: userIds,
    stand,
    registration_count: registrationCount ?? (stand === 1 ? 1 : 0),
  })

export const cancelRegistration = (activityId: string, userId: number) =>
  http.delete<void>(`/activities/${activityId}/user/${userId}/registration`)
